rfc/composite_table_availability.pseudocode

```diff
@@ -0,0 +1,123 @@
+
+PhysicalTableSchema:
+	getFieldName(Column c)
+
+DataSource:
+	String getName()
+	Map<FieldName, Set<Segment>> raw segment data
+	Map<FieldName, List<Interval>> squashedFromSegments  // optional optimization
+
+	Set<FieldName> getFields() //union of all keys
+	List<Interval> getIntervals(FieldName fieldName)
+	Map<FieldName,List<Interval>> getIntervals(Set<FieldName>) // map slice from squashedMap
+	Set<SegmentId> getSegments(List<Interval>)  // Filter all segments by overlap on time
+
+DataSourceConstraints // Simple bean parameter object
+
+	getGrain
+	Set<FieldName> getMetrics
+	Set<Dimension> getDimensions
+	getApiFilters
+
+	DataSourceConstraints(ApiRequest)
+	DataSourceConstraints (DruidQuery, ApiFilter apifilters)
+
+	withMetrics(Set<FieldName> metrics) // Copy constructor
+
+
+DataSourceAvailabilityResolver:
+	List<String> getDataSourceNames(DataSourceConstraints)
```

Comment:
I can sub in 'DataSourceName' as a class, but to the best of my intuition, all the uses are as an identity and for connecting to serialized usages.